File: tools/gerar-licenca.js
#!/usr/bin/env node
// Ferramenta interna (offline) da NS Sistemas para gerar a Chave de Ativação
// (RSA) enviada a um comprador do Mercado Livre — NÃO faz parte do app
// distribuído. Requer `src-tauri/licensing/private_key.pem` (gerado por
// `cargo run --bin gerar_chaves`, nunca versionado). Equivalente em Node.js
// de `src-tauri/src/bin/gerar_licenca.rs` — mesma assinatura RSASSA-PKCS1-v1_5
// com SHA-256 sobre o HWID, então as duas ferramentas produzem chaves idênticas.
//
// Uso:
//   node tools/gerar-licenca.js --hwid NS-XXXX-XXXX-XXXX
//
// O HWID vem da tela de Ativação de Licença do cliente (TelaAtivacaoLicenca.jsx,
// botão "Copiar HWID"). A chave impressa é o que o comprador cola no campo
// "Chave de Ativação (RSA)" daquela tela.

import { createSign } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PRIVATE_KEY_PATH = join(__dirname, "..", "src-tauri", "licensing", "private_key.pem");

function parseHwid(argv) {
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--hwid") return argv[i + 1];
    if (argv[i].startsWith("--hwid=")) return argv[i].slice("--hwid=".length);
  }
  return undefined;
}

const hwid = parseHwid(process.argv.slice(2))?.trim();

if (!hwid) {
  console.error("Uso: node tools/gerar-licenca.js --hwid NS-XXXX-XXXX-XXXX");
  process.exit(1);
}

let privateKeyPem;
try {
  privateKeyPem = readFileSync(PRIVATE_KEY_PATH, "utf8");
} catch {
  console.error(
    `Não encontrei ${PRIVATE_KEY_PATH}. Rode primeiro "cargo run --bin gerar_chaves" (dentro de src-tauri).`
  );
  process.exit(1);
}

const chave = createSign("RSA-SHA256").update(hwid).end().sign(privateKeyPem, "base64");

console.log(`HWID: ${hwid}`);
console.log(`Chave de Ativação (RSA): ${chave}`);

File: src-tauri/src/bin/gerar_licenca.rs
// Ferramenta interna (offline) da NS Sistemas para gerar a Chave de Ativação
// (RSA) enviada a um cliente — NÃO faz parte do app distribuído. Requer
// `licensing/private_key.pem` (gerado por `gerar_chaves`, nunca versionado).
//
// Uso, a partir de `src-tauri`:
//   cargo run --bin gerar_licenca -- NS-XXXX-XXXX-XXXX
//
// O HWID vem da tela de Ativação de Licença do cliente (TelaAtivacaoLicenca.jsx,
// link discreto de suporte no rodapé). A chave impressa é o que o cliente cola
// no campo "Chave de Licença Serial" daquela tela.
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use rsa::pkcs8::DecodePrivateKey;
use rsa::{Pkcs1v15Sign, RsaPrivateKey};
use sha2::{Digest, Sha256};
use std::fs;

fn main() {
    let hwid = std::env::args().nth(1).unwrap_or_default();
    let hwid = hwid.trim();
    if hwid.is_empty() {
        eprintln!("Uso: cargo run --bin gerar_licenca -- NS-XXXX-XXXX-XXXX");
        std::process::exit(1);
    }

    let private_pem = fs::read_to_string("licensing/private_key.pem").unwrap_or_else(|_| {
        eprintln!(
            "Não encontrei licensing/private_key.pem. Rode primeiro `cargo run --bin gerar_chaves`."
        );
        std::process::exit(1);
    });
    let private_key =
        RsaPrivateKey::from_pkcs8_pem(&private_pem).expect("chave privada inválida em licensing/private_key.pem");

    let mut hasher = Sha256::new();
    hasher.update(hwid.as_bytes());
    let hashed = hasher.finalize();

    let signature = private_key
        .sign(Pkcs1v15Sign::new::<Sha256>(), &hashed)
        .expect("falha ao assinar o HWID");

    let chave = BASE64.encode(signature);
    println!("HWID: {hwid}");
    println!("Chave de Ativação (RSA): {chave}");
}
